### PR TITLE
[MPS] Use metal shaders for all view ops

### LINKED ISSUE #143140: Crash with complex64 tensors and einsum on mps
### 🐛 Describe the bug

I'm working with complex valued tensors on mps. 

In my network I have a function that simply does:

`weights = torch.view_as_complex(weights)`
`return torch.einsum("bixyz,ioxyz->boxyz", Input, weights)`

Input is shape 1,16,24,24,24, complex64
weights is shape 16,24,24,24 (after torch.view_as_complex) complex64

This crashes Python, with the following message:

(mpsFileLoc): /AppleInternal/Library/BuildRoots/4b66fb3c-7dd0-11ef-b4fb-4a83e32a47e1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:137:0: error: 'mps.gather' op operand #0 must be tensor of mps native type values, but got 'tensor<2162688xcomplex<f32>>'
(mpsFileLoc): /AppleInternal/Library/BuildRoots/4b66fb3c-7dd0-11ef-b4fb-4a83e32a47e1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:137:0: note: see current operation: %38 = "mps.gather"(%34, %36, %37) <{batch_dims = 0 : ui32}> : (tensor<2162688xcomplex<f32>>, tensor<221184xsi32>, tensor<si32>) -> tensor<221184xcomplex<f32>>
(mpsFileLoc): /AppleInternal/Library/BuildRoots/4b66fb3c-7dd0-11ef-b4fb-4a83e32a47e1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:137:0: error: 'mps.gather' op operand #0 must be tensor of mps native type values, but got 'tensor<2162688xcomplex<f32>>'
(mpsFileLoc): /AppleInternal/Library/BuildRoots/4b66fb3c-7dd0-11ef-b4fb-4a83e32a47e1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphUtilities.mm:137:0: note: see current operation: %38 = "mps.gather"(%34, %36, %37) <{batch_dims = 0 : ui32}> : (tensor<2162688xcomplex<f32>>, tensor<221184xsi32>, tensor<si32>) -> tensor<221184xcomplex<f32>>
/AppleInternal/Library/BuildRoots/4b66fb3c-7dd0-11ef-b4fb-4a83e32a47e1/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraphExecutable.mm:988: failed assertion `original module failed verification'
/var/folders/p9/y3pw6wws17z1vf8r12lv3ml40000gn/T/spyder-james/tmpdkad8_ea: line 3:  1508 Abort trap: 6           /opt/anaconda3/envs/pytorch/bin/python -Xfrozen_modules=off -m spyder_kernels.console -f /Users/james/Library/Jupyter/runtime/kernel-d94981e9deec.json

The kernel died, restarting...
 

Fatal Python error: Aborted

Main thread:
Current thread 0x00000001ff8cb840 (most recent call first):
  File "/opt/anaconda3/envs/pytorch/lib/python3.12/site-packages/torch/functional.py", line 402 in einsum

### Versions

PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.1.1 (arm64)
GCC version: Could not collect
Clang version: 16.0.0 (clang-1600.0.26.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.7 | packaged by Anaconda, Inc. | (main, Oct  4 2024, 08:22:19) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-15.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3

Versions of relevant libraries:
[pip3] configmypy==0.2.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.0.1
[pip3] tensorly-torch==0.5.0
[pip3] torch==2.5.1
[pip3] torchaudio==2.5.1
[pip3] torchvision==0.20.1
[conda] numpy                     2.0.1           py312h7f4fdc5_1  
[conda] numpy-base                2.0.1           py312he047099_1  
[conda] pytorch                   2.5.1                  py3.12_0    pytorch
[conda] tensorly-torch            0.5.0                    pypi_0    pypi
[conda] torchaudio                2.5.1                 py312_cpu    pytorch
[conda] torchvision               0.20.1                py312_cpu    pytorch

cc @ezyang @anjali411 @dylanbespalko @mruberry @nikitaved @amjames @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by malfet**
Before this PR Metal  shaders were used to scatter/gather 1-5 dimensional tensors. 
This PR introduces generalized ones that could be used for any dimensionality and as results  gets rid of 700+ lines complex and untested code that might not even work as expected.
Generalized gather shader looks as follows
```metal
kernel void gather_kernel_n(uint linear_index           [[thread_position_in_grid]],
                            constant void * src_        [[buffer(0)]],
                            device void * dst_          [[buffer(1)]],
                            constant uint32_t * size    [[buffer(2)]],
                            constant uint32_t * stride  [[buffer(3)]],
                            constant uint32_t & numel   [[buffer(4)]],
                            constant int32_t & ndim     [[buffer(5)]]) {{
    if (linear_index >= numel) return;

    constant {0} * src = (constant {0} *)src_;
    device {1} * dst = (device {1} *)dst_;


    uint64_t src_offs = 0;
    auto src_idx = linear_index;
    for(int dim = ndim - 1; dim >= 0; --dim) {{
      src_offs += stride[dim] * (src_idx % size[dim]);
      src_idx /= size[dim];
    }}

    dst[linear_index] = cast<{1}>(src[src_offs]);
}}
```

Which, according to the following benchmark
```python
from timeit import default_timer

import torch
import torch.utils.cpp_extension
from torch.utils.benchmark import Measurement, Timer

t = Timer(
    stmt=f"y.copy_(x);torch.mps.synchronize()",
    setup=f"x=torch.rand(4, 5, 16, 64, 33, 24, dtype=torch.float32, device='mps')[:,:,:,:24,:24,];y=torch.empty(x.shape, device=x.device, dtype=x.dtype)",
    language="python", timer=default_timer
)
print(t.blocked_autorange())
```
Is almost twice as fast as previous implementation (i.e. on Mac Book M2 Pro it returns 2.9ms for MPS version vs 1.5ms for shader one

On MacOS Sequoia [`gatherWithUpdatesTensor: indicesTensor:...`](https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/gather(withupdatestensor:indicestensor:axis:batchdimensions:name:)?language=objc) crashes if invoked with complex data type, as one can see by running the code below
```swift
import Metal
import MetalPerformanceShadersGraph


func gatherComplexMPS(device: MTLDevice,
                inp_buf: MTLBuffer, idx_buf: MTLBuffer,
                out_buf: MTLBuffer,
                inp_elem: Int, upd_elem: Int) {
  let graph = MPSGraph()
  let inputPlaceholder = graph.placeholder(shape: [inp_elem as NSNumber], dataType: .complexFloat32, name: nil)
  let indicesPlaceholder = graph.placeholder(shape: [upd_elem as NSNumber], dataType: .int64, name: nil)
  let outNode = graph.gather(withUpdatesTensor: inputPlaceholder, indicesTensor: indicesPlaceholder, axis: 0, batchDimensions: 0, name: nil)
  let mpsInputBuffer = MPSGraphTensorData(inp_buf, shape: [inp_elem as NSNumber], dataType: .complexFloat32)
  let mpsIndicesBuffer = MPSGraphTensorData(idx_buf, shape: [upd_elem as NSNumber], dataType: .int64)
  let mpsOutputBuffer = MPSGraphTensorData(out_buf, shape: [inp_elem as NSNumber], dataType: .complexFloat32)
  guard let queue = device.makeCommandQueue() else { fatalError("Can't make queue") }
  graph.run(with: queue, feeds: [inputPlaceholder: mpsInputBuffer,
                               indicesPlaceholder: mpsIndicesBuffer ],
            targetOperations: nil, resultsDictionary: [outNode: mpsOutputBuffer])
}

func makeBufferWithValues<T>(device: MTLDevice, values: [T]) -> MTLBuffer {
  guard let buf = device.makeBuffer(length: values.count * MemoryLayout<T>.size, options: [.storageModeShared]) else { fatalError("Can't alloc") }
  let buf_data = buf.contents().assumingMemoryBound(to: T.self)
  for i in 0..<values.count {
    buf_data[i] = values[i]
  }
  return buf
}

guard let device = MTLCopyAllDevices().first else { fatalError("Not Metal device found") }
print("Using device \(device.name)")

let inp_buf = makeBufferWithValues(device: device, values: [1.0, 2.0 , 3.0, 4.0, 5.0, 6.0, 7.0, 8.0])
let idx_buf = makeBufferWithValues(device: device, values: [0, 1, 2, 3])
guard let out_buf = device.makeBuffer(length:8 * MemoryLayout<Float>.size, options: [.storageModeShared]) else { fatalError("Can't alloc") }

gatherComplexMPS(device: device, inp_buf: inp_buf, idx_buf: idx_buf, out_buf: out_buf, inp_elem: 4, upd_elem: 4)
```

Fixes https://github.com/pytorch/pytorch/issues/143140

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/View.mm [hunk 1]
```diff
@@ -666,69 +38,9 @@ static IntArrayRef updateTensorBaseShape(const Tensor& self) {
...
-      newCachedGraph->outputTensor =
-          chainViewOperation(newCachedGraph, size, stride, storage_offset, base_shape, needsScatter, updatesTensor);
-    });
-  }
-}
-
 static std::string getGatherScatterFunctionName(ScalarType scalarType, int64_t dim, bool needsScatter) {
   std::string kernelName = needsScatter ? "scatter" : "gather";
-  return kernelName + "_kernel_" + std::to_string(dim == 0 ? 1 : dim);
+  return kernelName + "_kernel_" + (dim < 5 ? std::to_string(dim == 0 ? 1 : dim) : "n");
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/View.mm:43]: Is the idea here that we have special case for 1-4 dimensions and then a shared kernel for 5+?
The reason we do that is that the 1-4 have explicit args which is faster to pass in while the 5+ is pointer-based passed and so more expensive?
This would be worth a small comment maybe?

Also what's the max number of dimensions accepted here? Do we test for it?

File: aten/src/ATen/mps/IndexKernels.h [hunk 2]
```diff
@@ -15,32 +11,26 @@ template<>
...
-    packed_uint5 strided_index;
-    strided_index.x = local_index.x * stride.x;
-    strided_index.y = local_index.y * stride.y;
-    strided_index.z = local_index.z * stride.z;
-    strided_index.w = local_index.w * stride.w;
-    strided_index.u = local_index.u * stride.u;
-
-    dst[strided_index.x + strided_index.y + strided_index.z + strided_index.w + strided_index.u] = cast<{1}>(src[linear_index]);
+    uint64_t dst_offs = 0;
+    uint dst_idx = linear_index;
```
Comment (albanD) [inline aten/src/ATen/mps/IndexKernels.h:27]: should we specific width for all integer types? tbh I have no clue how wide an "int" is anymore :D

File: aten/src/ATen/mps/IndexKernels.h [hunk 3]
```diff
@@ -15,32 +11,26 @@ template<>
...
+    for(int dim = ndim - 1; dim >= 0; --dim) {{
```
Comment (albanD) [inline aten/src/ATen/mps/IndexKernels.h:28]: Any perf hit to worry about?
Would a template + unrolling be important here?

**Review by albanD — APPROVED** [review]
nit on docs and details, but SGTM once these are cleaned up

File: aten/src/ATen/native/mps/operations/View.mm (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/mps/operations/View.mm:43]: Let me leave a comment. 
I don't think passing args using packed_int makes kernel faster, but rather explicit loop unrolling and lack of boundary checks. I.e. even on cpu `int sum = 0; for(int i = 0; i < n; ++i) sum = v[i];` is 3x slower than `sum = v[0] + v[1];` if one can infer at compile time that `n` is 2

File: aten/src/ATen/mps/IndexKernels.h (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/mps/IndexKernels.h:27]: int is 32-bit. And yes, those kernels need to have a generic 64 bit variant, but this will be done in a followup PRs

File: aten/src/ATen/mps/IndexKernels.h (hunk 3 above)
Comment (malfet) [reply aten/src/ATen/mps/IndexKernels.h:28]: It'll be faster than MPS kernel, as this one required an intermediate tensor of all elements, will write a benchmark and attach it to PR description
